clog/settings.go

```diff
@@ -0,0 +1,172 @@
+	LevelDisabled logLevel = "disabled"
+)
+
+type logFormat string
+
+const (
+	// use for cli/terminal
+	FormatForHumans logFormat = "human"
+	// use for cloud logging
+	FormatToJSON logFormat = "json"
+)
+
+type sensitiveInfoHandlingAlgo string
+
+const (
+	HashSensitiveInfo            sensitiveInfoHandlingAlgo = "hash"
+	MaskSensitiveInfo            sensitiveInfoHandlingAlgo = "mask"
+	ShowSensitiveInfoInPlainText sensitiveInfoHandlingAlgo = "plaintext"
+)
+
+const (
+	Stderr = "stderr"
+	Stdout = "stdout"
+)
+
+// ---------------------------------------------------
+// configuration
+// ---------------------------------------------------
+
+// Default location for writing log files.
+var defaultLogFileDir = filepath.Join(os.Getenv("HOME"), "Library", "Logs")
+
+// ResolvedLogFile is the first log file established by the caller.
+// It gets eagerly populated on the first act of ensuring settings
+// defaults, which normally occurs during the Init call.
+//
+// If Init gets called more than once, or different settings are
+// ensured, it's possible to override this value by manually specifying
+// the log file in the settings used for that action.  But if no file
+// is provided, the default will fall back to this resolved file first.
+var ResolvedLogFile string
+
+// Settings records the user's preferred logging settings.
+type Settings struct {
+	// core settings
+	File   string    // what file to log to (alt: stderr, stdout)
+	Format logFormat // whether to format as text (console) or json (cloud)
+	Level  logLevel  // what level to log at
+
+	// more fiddly bits
+	SensitiveInfoHandling sensitiveInfoHandlingAlgo // how to obscure pii
+	// when non-empty, only debuglogs with a label that matches
+	// the provided labels will get delivered.  All other debug
+	// logs get dropped.  Good way to expose a little bit of debug
+	// logs without flooding your system.
+	OnlyLogDebugIfContainsLabel []string
+}
+
+// EnsureDefaults sets any non-populated settings to their default value.
+// exported for testing without circular dependencies.
+func (s Settings) EnsureDefaults() Settings {
+	set := s
+
+	levels := []logLevel{LevelDisabled, LevelDebug, LevelInfo, LevelError}
+	if len(set.Level) == 0 || !slices.Contains(levels, set.Level) {
+		set.Level = LevelInfo
+	}
+
+	formats := []logFormat{FormatForHumans, FormatToJSON}
+	if len(set.Format) == 0 || !slices.Contains(formats, set.Format) {
+		set.Format = FormatForHumans
+	}
+
+	algs := []sensitiveInfoHandlingAlgo{ShowSensitiveInfoInPlainText, MaskSensitiveInfo, HashSensitiveInfo}
+	if len(set.SensitiveInfoHandling) == 0 || !slices.Contains(algs, set.SensitiveInfoHandling) {
+		set.SensitiveInfoHandling = ShowSensitiveInfoInPlainText
+	}
+
+	if len(set.File) == 0 {
+		set.File = GetLogFileOrDefault("")
+	}
+
+	if len(ResolvedLogFile) == 0 {
+		ResolvedLogFile = set.File
+	}
+
+	return set
+}
+
+// Returns the default location for log file storage.
+func defaultLogLocation() string {
```

Comment:
setting the default location and using a global env var to set the location seem problematic. Now two programs that consume this package can't run on the same machine with the defaults, and it's not clear what could happen if people do run with the defaults

Reply:
I'm still rather uncomfortable with the assumptions being made about setup of the logging component. This is technically an open-source library and I'm sure we'd like to see usage in our own systems going forward. However, the logging setup has hidden configurations that will lead to difficult to debug issues if not noticed while also making some pretty heavy assumptions about usage in general (e.x. assume sent to stderr which isn't really the standard when it comes to outputting log files AFAIK)

If the goal is adoption and usage then, given that this is basically a core feature that everything is going to rely on, the implementation should be robust enough that folks can easily do what they need and are unlikely to hit unexpected issues

Reply:
I have no disagreement with your assesment.  But is it really a blocker to punt on the issue for this PR?  The code already matches this in the existing clog repo, so we're not _avoiding_ the issue at this moment.  And I will follow up on it.  But this PR is supposed to be a copy-paste, not a fixup of the repo.

Reply:
see: https://github.com/alcionai/clues/pull/68